zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -352,35 +354,29 @@ public String toString() {
      * established until needed. The start() instance method must be called
      * subsequent to construction.
      *
-     * @param chrootPath - the chroot of this client. Should be removed from this Class in ZOOKEEPER-838
-     * @param hostProvider
-     *                the list of ZooKeeper servers to connect to
-     * @param sessionTimeout
-     *                the timeout for connections.
-     * @param zooKeeper
-     *                the zookeeper object that this connection is related to.
-     * @param watcher watcher for this connection
-     * @param clientCnxnSocket
-     *                the socket implementation used (e.g. NIO/Netty)
-     * @param canBeReadOnly
-     *                whether the connection is allowed to go to read-only
-     *                mode in case of partitioning
-     * @throws IOException
+     * @param chrootPath the chroot of this client. Should be removed from this Class in ZOOKEEPER-838
+     * @param hostProvider the list of ZooKeeper servers to connect to
+     * @param sessionTimeout the timeout for connections.
+     * @param clientConfig the client configuration.
+     * @param defaultWatcher default watcher for this connection
+     * @param clientCnxnSocket the socket implementation used (e.g. NIO/Netty)
+     * @param canBeReadOnly whether the connection is allowed to go to read-only mode in case of partitioning
      */
     public ClientCnxn(
         String chrootPath,
         HostProvider hostProvider,
         int sessionTimeout,
-        ZooKeeper zooKeeper,
-        ClientWatchManager watcher,
+        ZKClientConfig clientConfig,
```

Comment:
this is my major concern.

It is true that we are not changing the very high level API, as ClientCnxn is an internal class and applications should not use it directly.

But it is probable that frameworks like Curator (ping @Randgalt ) do tricks over this stuff.

This change is a big refactor and touches the very internals of the client so we should handle it with care.

Don't get my wrong, I think this work is awesome and I hope we can make it land it to master someday.

That's only my opinion, let's see if any other reviewer wants to sponsor this change

Reply:
On first glance it doesn't look like it conflicts with Curator. But, maybe @TisonKun can test with Curator locally. As an aside, I wonder if we shouldn't include some Curator test in ZooKeeper in the future - i.e. validate that new changes don't break Curator.

Reply:
@eolivelli After a close look I find that where Curator hacking into `CliengCnxn` is only `InjectSessionExpiration` that get access of `cnxn`/`eventThread`/`queueEvent`. So this patch should not break Curator.

And I second @Randgalt that it is Curator responsibility to adjust its trick for accessing ZooKeeper internal concept since they don't promise compatibility. IMO Curator already has a `Compatibility` utils for different manner between zk34 and zk35. We should continue refining codebase so that projects such as Curator don't have to rely on internal concepts, though.